sprint-3

sgp900 opened the pull request:
- Добавлен роутинг;
- В модели MVC, добавлены контроллеры и модели;
- Работа с сервером через HTTP | WebSocket;
- Реализована функциональность:
  1. Вход и Регистрация нового пользователя, а также выход из приложения
  2. Возможность изменять основную информацию о пользователе, а также его аватар и пароль
  3. Создавать новые чаты
  4. Читать старые и писать новые сообщения в чат
  5. Добавлять и удалять пользователя в чат

sgp900 committed:
Sprint 3

- Добавлен роутинг;
- В модели MVC, добавлены контроллеры и модели;
- Работа с сервером через HTTP | WebSocket;
- Реализована функциональность:
  1. Вход и Регистрация нового пользователя, а также выход из приложения
  2. Возможность изменять основную информацию о пользователе, а также его аватар и пароль
  3. Создавать новые чаты
  4. Читать старые и писать новые сообщения в чат
  5. Добавлять и удалять пользователя в чат

gennady-bars commented on src/controllers/UserController.ts:
```diff
@@ -0,0 +1,60 @@
+      await req();
+      store.set("profile.error", undefined);
+    } catch (e: any) {
+      store.set("profile.error", e.reason);
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/utils/WSTransport.ts:
```diff
@@ -0,0 +1,65 @@
+
+  private subscribe(socket: WebSocket) {
+    socket.addEventListener("message", (message) => {
+      const data = JSON.parse(message.data);
```
`JSON.parse` может выкинуть исключение, если данные **невалидные**. Стоит добавить блок `try/catch` для отлова возможной ошибки

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -1,86 +1,112 @@
 
-type HTTPMethod = (url: string, options?: Options) => Promise<unknown>;
+class HTTPTransport {
+  static API_URL = "https://ya-praktikum.tech/api/v2";
```
Отлично, что базовый урл вынесен в константу

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -1,86 +1,112 @@
 
-  delete: HTTPMethod = (url, options = {}) =>
-    this.request(url, { ...options, method: METHODS.DELETE }, options.timeout);
+  public post<Response = void>(
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod<Response = void> = (url: string, options?: Options) => Promise<Response>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* `Readme` хорошо оформлен
* Проект задеплоен и отлично работает
* Отлично, что фиксируете версии зависимостей
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Хороший роутинг
* Хорошая логика сокетов
* Отлично, что базовый урл вынесен в константу


но есть некоторые недочеты: 

* уже не нужны ссылки на 404 и 500 страницы. Скрин  https://disk.yandex.ru/i/bhnYRSVDKvMQHQ . Они нужны были только на 1-2 спринтах
* когда перезагружаю страницу, находясь в чатах, то попадаю на страницу 404 от Нетлифай, а роутинг должен отправлять пользователя по эндпоинту, который указан в урле
* должна быть возможность удалить чат
* `JSON.parse` может выкинуть исключение, если данные **невалидные**. Стоит добавить блок `try/catch` для отлова возможной ошибки





**Можно лучше**

* Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище
* Еще должна быть возможность добавлять картинку для чата (аватарку). Предусмотрен такой функционал
* Попробуйте избавиться от `any`




Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

sgp900 committed:
1. Удалены ссылки на страницы 404 и 500;
2. В конфигурационный файл Netlify добавлены свойства для редиректа;
3. Добавлена возможность удалять чаты;
4. Отлавливаем ошибки при вызове JSON.parse;
5. Типизированы функции post, get, put etc в HTTPTransport;
6. Добавлена возможность добавлять аватарку для чатов;

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
